src/datachain/data_storage/warehouse.py

```diff
@@ -227,7 +227,10 @@ def dataset_select_paginated(
                 if limit < page_size:
                     paginated_query = paginated_query.limit(None).limit(limit)
 
-            results = self.dataset_rows_select(paginated_query.offset(offset))
+            # Ensure we're using a thread-local connection
+            with self.clone() as wh:
+                # Cursor results are not thread-safe, so we convert them to a list
```

Comment:
Q: why do they have to be thread safe (I mean cursor results)?
Q: are there any implications in terms of memory usage for this?

Reply:
With sqlite, the cursor holds all the results in memory anyway, so the conversion shouldn't change things much, though we do now temporarily have 2 copies of the full result set.
I'm not sure about the behaviour of the ClickHouse driver, OTOH. But in any case, pagination mitigates the memory issues, as we're dealing with at most 100k rows at a time.